File: mobile/src/main/java/technology/mainthread/apps/gatekeeper/view/fragment/SettingsFragment.java
package technology.mainthread.apps.gatekeeper.view.fragment;

import android.os.Bundle;
import android.os.Vibrator;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.view.View;
import android.widget.Toast;

import javax.inject.Inject;

import de.psdev.licensesdialog.LicensesDialog;
import technology.mainthread.apps.gatekeeper.GatekeeperApp;
import technology.mainthread.apps.gatekeeper.R;
import technology.mainthread.apps.gatekeeper.common.rx.RxSchedulerHelper;
import technology.mainthread.apps.gatekeeper.data.AndroidAppInfo;
import technology.mainthread.apps.gatekeeper.data.AuthManager;
import technology.mainthread.apps.gatekeeper.data.VibratorTunes;
import technology.mainthread.apps.gatekeeper.view.activity.AuthActivity;

public class SettingsFragment extends RxPreferenceFragment {

    @Inject
    AndroidAppInfo androidAppInfo;
    @Inject
    Vibrator vibrator;
    @Inject
    AuthManager authManager;

    public static Fragment newInstance() {
        return new SettingsFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        GatekeeperApp.get(getActivity()).inject(this);
    }

    @Override
    public void onCreatePreferences(Bundle bundle, String s) {
        addPreferencesFromResource(R.xml.preferences);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        findPreference("pref_sign_out").setOnPreferenceClickListener(preference -> {
            authManager.signOut()
                    .compose(RxSchedulerHelper.applyFlowableSchedulers())
                    .compose(bindToLifecycle())
                    .subscribe(success -> {
                        if (success) {
                            showToast(R.string.sign_out_success);
                            onSignOutSuccess();
                        } else {
                            showToast(R.string.sign_out_failure);
                        }
                    });
            return true;
        });
        findPreference("pref_delete").setOnPreferenceClickListener(preference -> {
            authManager.deleteAccount()
                    .compose(RxSchedulerHelper.applyFlowableSchedulers())
                    .compose(bindToLifecycle())
                    .subscribe(success -> {
                        if (success) {
                            showToast(R.string.delete_account_success);
                            onSignOutSuccess();
                        } else {
                            showToast(R.string.delete_account_failure);
                        }
                    });
            return true;
        });

        findPreference("pref_build").setSummary(String.format("%1$s (%2$s)", androidAppInfo.getVersionName(), androidAppInfo.getVersionCode()));
        findPreference("pref_licenses").setOnPreferenceClickListener(preference -> {
            showLicencesDialog();
            return true;
        });

        findPreference("pref_vibrate_tune").setOnPreferenceChangeListener((preference, newValue) -> {
            long[] vibrateTune = VibratorTunes.VIBRATE_TUNES.get(newValue);
            if (vibrateTune != null && vibrateTune.length > 0) {
                vibrator.vibrate(vibrateTune, -1);
            }

            return true;
        });
    }

    private void showLicencesDialog() {
        if (!isRemoving()) {
            new LicensesDialog.Builder(getActivity()).setNotices(R.raw.notices).setIncludeOwnLicense(true).build().show();
        }
    }

    private void showToast(@StringRes int message) {
        Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
    }

    private void onSignOutSuccess() {
        getActivity().startActivity(AuthActivity.getIntent(getActivity()));
        getActivity().finish();
    }

}

File: mobile/src/main/java/technology/mainthread/apps/gatekeeper/view/fragment/SettingsFragment.kt
package technology.mainthread.apps.gatekeeper.view.fragment

import android.os.Bundle
import android.os.Vibrator
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import android.view.View
import android.widget.Toast
import de.psdev.licensesdialog.LicensesDialog
import io.reactivex.disposables.CompositeDisposable
import technology.mainthread.apps.gatekeeper.R
import technology.mainthread.apps.gatekeeper.rx.applyFlowableSchedulers
import technology.mainthread.apps.gatekeeper.data.AndroidAppInfo
import technology.mainthread.apps.gatekeeper.data.AuthManager
import technology.mainthread.apps.gatekeeper.data.VibratorTunes
import technology.mainthread.apps.gatekeeper.rx.plusAssign
import technology.mainthread.apps.gatekeeper.view.activity.getAuthIntent
import technology.mainthread.apps.gatekeeper.view.baseHelpers.DaggerPreferenceFragment
import javax.inject.Inject

fun buildSettingsFragment(): Fragment {
    return SettingsFragment()
}

class SettingsFragment : DaggerPreferenceFragment() {

    @Inject
    internal lateinit var androidAppInfo: AndroidAppInfo
    @Inject
    internal lateinit var vibrator: Vibrator
    @Inject
    internal lateinit var authManager: AuthManager

    private var disposables: CompositeDisposable = CompositeDisposable()

    override fun onCreatePreferences(bundle: Bundle?, s: String?) {
        addPreferencesFromResource(R.xml.preferences)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        findPreference("pref_sign_out").setOnPreferenceClickListener { _ ->
            disposables += authManager.signOut()
                    .compose(applyFlowableSchedulers<Boolean>())
                    .subscribe { success ->
                        if (success) {
                            showToast(R.string.sign_out_success)
                            onSignOutSuccess()
                        } else {
                            showToast(R.string.sign_out_failure)
                        }
                    }
            true
        }
        findPreference("pref_delete").setOnPreferenceClickListener { _ ->
            disposables += authManager.deleteAccount()
                    .compose(applyFlowableSchedulers<Boolean>())
                    .subscribe { success ->
                        if (success) {
                            showToast(R.string.delete_account_success)
                            onSignOutSuccess()
                        } else {
                            showToast(R.string.delete_account_failure)
                        }
                    }
            true
        }

        findPreference("pref_build").summary = String.format("%1\$s (%2\$s)", androidAppInfo.versionName, androidAppInfo.versionCode)
        findPreference("pref_licenses").setOnPreferenceClickListener { _ ->
            showLicencesDialog()
            true
        }

        findPreference("pref_vibrate_tune").setOnPreferenceChangeListener { _, newValue ->
            val vibrateTune = VibratorTunes.VIBRATE_TUNES[newValue]
            if (vibrateTune != null && vibrateTune.isNotEmpty()) {
                vibrator.vibrate(vibrateTune, -1)
            }

            true
        }
    }

    override fun onDestroyView() {
        disposables.dispose()
        super.onDestroyView()
    }

    private fun showLicencesDialog() {
        if (!isRemoving) {
            LicensesDialog.Builder(activity).setNotices(R.raw.notices).setIncludeOwnLicense(true).build().show()
        }
    }

    private fun showToast(@StringRes message: Int) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    private fun onSignOutSuccess() {
        activity?.startActivity(getAuthIntent(activity!!))
        activity?.finish()
    }

}
